Fix CSE to not construct invalid tree

jlherren opened the pull request:
Without executing Mul's post-processor, it might pack matrices inside.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #18991

#### Other comments
This raises the question whether maybe all `_from_args()` methods should themselves invoke the post processors instead, to avoid having to do that from all call sites. Existing code seems to call it after `_from_args()` though.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

jlherren commented:
This travis failure seems unrelated... The exact same error happens also on other pull requests.

oscarbenjamin commented:
See #19084

jlherren commented:
Yes, post-processors are evaluated even with `evaluate=False` as can be seen by executing `type(Mul(MatrixSymbol('A', 2, 2), MatrixSymbol('B', 2, 2), evaluate=False))` which evaluates to `MatMul`.

asmeurer commented:
So I think we should use that here instead. `_from_args` should only be used internally in the core.

jlherren committed:
Fix CSE to not construct invalid tree

Cannot use Mul._from_args() here because it fails to return MatMul if an
argument is a matrix.

jlherren commented:
That does make sense, so now I just call `Mul(...)`.

asmeurer commented on sympy/simplify/cse_opts.py:
```diff
@@ -37,7 +38,7 @@ def sub_post(e):
         if isinstance(node, Mul) and \
             node.args[0] is S.One and node.args[1] is S.NegativeOne:
-            replacements.append((node, -Mul._from_args(node.args[2:])))
+            replacements.append((node, -Mul(*node.args[2:])))
```
I think this one should also be evaluate=False. _from_args doesn't evaluate.

jlherren replied:
Ah, stupid me, thanks.

asmeurer commented:
That looks better, except for the one comment. Since the tests pass, I think this fix is OK.
